### PR TITLE
Add HNS Load Balancer Healthchecks for ExternalTrafficPolicy: Local

### LINKED ISSUE #62046: Local traffic policy not working on Windows
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
/kind bug
> /kind feature

**What happened**:

Also onlyNodeLocalEndpoints is handled when processing service updates in `pkg/proxy/winkernel`, it is not used in syncProxyRules.

We should fix this and cherry-pick to old releases.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by anfernee**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it:**
In GCE, the current `externalTrafficPolicy: Local` logic does not work because the nodes that run the pods do not setup load balancer ports. This means that the GCLB does not understand which nodes are serving the pods that can accept traffic. Since all report unhealthy it'll direct traffic to any node. This PR configures the health check ports so that GCLB knows which nodes can handle the traffic.

See #62046 for details.
This work is based on #96998 

**Which issue(s) this PR fixes:**

Fixes #62046

**Special notes for your reviewer:**

Does this PR introduce a user-facing change?:

`externalTrafficPolicy: local` works on Windows hosts with DSR support.

**Release Note**
```release-note
Add 2 new options for kube-proxy running in winkernel mode. 
`--forward-healthcheck-vip`, if specified as true, health check traffic whose destination is service VIP will be forwarded to kube-proxy's healthcheck service. `--root-hnsendpoint-name` specifies the name of the hns endpoint for the root network namespace.
This option enables the pass-through load balancers like Google's GCLB to correctly health check the backend services. Without this change, the health check packets is dropped, and Windows node will be considered to be unhealthy by those load balancers.
```

### CODE REVIEW

**Comment by anfernee** [comment]
@madhanrm @elweb9858 PTAL.

File: pkg/proxy/winkernel/proxier.go
Comment (jeremyje) [inline pkg/proxy/winkernel/proxier.go:1290]: Should the flags here be synced with the flags above?

Comment (anfernee) [reply pkg/proxy/winkernel/proxier.go:1290]: Actually I am not sure what this annotation means here: `preserveDIP := service.Annotations["preserve-destination"] == "true"`, they are copied over.

The rest is not necessarily synced, because the only targeting client is GCLB health check, the requirement for the service itself doesn't apply to health check.

**Comment by anfernee** [comment]
friendly ping.. @andrewsykim @ksubrmnn

**Comment by daschott** [comment]
@elweb9858 @Keith-Mange PTAL

**Comment by elweb9858** [comment]
@sbangari could you take a look at this?

**Comment by anfernee** [comment]
@ibabou one more try to have this merged.

**Review by ibabou — APPROVED** [review]

**Comment by marosset** [comment]
@anfernee / @sbangari is this ready for review again?

**Comment by anfernee** [comment]
> @anfernee / @sbangari is this ready for review again?

Yes! I added 2 options to guard the behavior as @thockin suggested. Appreciated it if you can take a look.

**Comment by thockin** [comment]
Is this shooting for 23?

**Comment by anfernee** [comment]
> Is this shooting for 23?

Yes, it would be nice to include in.

**Comment by jyotimahapatra** [comment]
Hi 👋 I'm the bug triage shadow. We're in code freeze for 1.23. Just wanted to check if we can still get this PR merged in time for release(7/12)

**Comment by jyotimahapatra** [comment]
Hi 👋 I'm the bug triage shadow. We're in code freeze for 1.23 milestone. I wanted to check if we're targeting to get this PR merged in time for 1.23 release(7/12)

File: cmd/kube-proxy/app/server_windows.go
Comment (aojea) [inline cmd/kube-proxy/app/server_windows.go:105]: this seems is ok since is already validated in https://github.com/kubernetes/kubernetes/blob/a53e2eaeaba064309dceca2dc27f3ac09c6375b0/pkg/proxy/apis/config/validation/validation.go

File: pkg/proxy/winkernel/proxier.go
Comment (aojea) [inline pkg/proxy/winkernel/proxier.go:1326]: but this is not correct, this is about the healthz of the apiserver, not about if the service has local endpoints ... but to be honest I don't fully understand the problem, what is the TL; DR, of why the NewServiceHealthServer doesn't work in windows? that should fix the problem, no?
```
	serviceHealthServer := healthcheck.NewServiceHealthServer(hostname, recorder, []string{} /* windows listen to all node addresses */)
```

Comment (anfernee) [reply pkg/proxy/winkernel/proxier.go:1326]: The problem is it's not the node address. The google load balancer works this way:

```
User --> [clientip:12345 to lb_vip:80] -> Google LB -> [clientip:12345 to lb_vip:80] -> Node -> [clientip:12345 to pod_ip:80] -> Pod 
```
This works right now in kubeproxy's winkernel mode, because the rules are already in place.

What doesn't work is the health check for normal service:
```
Google LB -> [lb:12345 to lb_vip:10256] -> Node -> X(dropped)
```
and for `externalTrafficPolicy: local` service
```
Google LB -> [lb:12345 to lb_vip:<per-service-hc-port>] -> Node -> X(dropped)
```
They are dropped because the destination is not among any node IPs.

This basically adds hns rules to change request to node IP.

Comment (aojea) [reply pkg/proxy/winkernel/proxier.go:1326]: yeah, but why it uses the kube-proxy healthz port , it should use the healthzcheck nodeport, no?

Comment (aojea) [reply pkg/proxy/winkernel/proxier.go:1326]: the part I'm lost is about when it has to use healthz port, but I'm not familiar with the internals of GCE LBs, maybe we should get a reviewer more familiar with this :sweat_smile:

Comment (anfernee) [reply pkg/proxy/winkernel/proxier.go:1326]: > yeah, but why it uses the kube-proxy healthz port , it should use the healthzcheck nodeport, no?

`proxier.healthzPort` is basically 10256 if it's not explicitly configured: 

https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/cmd/kube-proxy/app/server.go#L212-L220

https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/cmd/kube-proxy/app/server.go#L226

https://github.com/kubernetes/kubernetes/blob/1bd00776b5d78828a065b5c21e7003accc308a06/pkg/cluster/ports/ports.go#L36

this is basically the healthzcheck nodeport. For `externalTrafficPolicy: local`, the port is different for every service. 

> the part I'm lost is about when it has to use healthz port, but I'm not familiar with the internals of GCE LBs, maybe we should get a reviewer more familiar with this 😅

I understand it's really not straight-forward.. I'll try to get someone from GKE service team to review as well :) @robscott @bowei can you guys help take a look?

Comment (robscott) [reply pkg/proxy/winkernel/proxier.go:1326]: Just catching up on this thread. I _think_ this makes sense. Although we don't need to do this in other proxier implementations, it makes sense that we'd need to this lb_vip -> node ip translation here to get health checks working. I'm less familiar with how the primary/default health check needs to work here. Regardless of OS, that seems to be handled exclusively at a higher level, with no extra logic added in either iptables or winkernel proxier implementations:

https://github.com/kubernetes/kubernetes/blob/481459d12dc82ab88e413886e2130c2a5e4a8ec4/cmd/kube-proxy/app/server_others.go#L153
https://github.com/kubernetes/kubernetes/blob/481459d12dc82ab88e413886e2130c2a5e4a8ec4/cmd/kube-proxy/app/server_windows.go#L101

I'm not familiar enough with how this is setup to know what's necessary, but is it possible that this works without any changes for the default health check? The original bug as reported by the PR seems to indicate the problem is only with `ExternalTrafficPolicy: Local`.

File: cmd/kube-proxy/app/init_windows.go
Comment (robscott) [inline cmd/kube-proxy/app/init_windows.go:44]: Nit: I think `HnsEndpoint` is a better way to represent this name throughout. That would also match the capitalization above in k8s-node-setup.psm1.

File: pkg/proxy/winkernel/hnsV1.go
Comment (robscott) [inline pkg/proxy/winkernel/hnsV1.go:121]: Nit: This TODO (and the other similar one) may fit a bit better on the type definition.

**Review by robscott — COMMENTED** [review]
Thanks for the work on this @anfernee!

File: pkg/proxy/winkernel/proxier.go
Comment (anfernee) [reply pkg/proxy/winkernel/proxier.go:1326]: at higher level, `healthcheck.NewProxierHealthServer()` is called to create http server that listens on the specific bind address for health check purpose. Because it listens on `0.0.0.0`, on Linux it will accept the connection regardless of destination. Unfortunately for Windows, the additional forward needs to be done to forward VIP to node-ip so the health check server can receive traffic.

**Comment by robscott** [comment]
Thanks @anfernee!

/remove-lifecycle stale
/triage accepted
/lgtm

**Comment by aojea** [comment]
/approve
/hold
for windows folks to take a look and unhold when they are happy
/assign @jsturtevant @jayunit100

**Comment by jayunit100** [comment]
this is tricky to reason about....   the urgency on merging this patch ? we could merge it now, watch it closely, and iterate where needed OR, we could ... 
- along with this patch,  add a `sig-windows` tagged test that ensures that `externalTrafficPolicy=local` when implemented, always sends out correct health checks?  i think it would be doable ? 
- otherwise, if you file an issue in https://github.com/K8sbykeshed/k8s-service-lb-validator/, our new service lb tests we can try to create this test for you .

**Comment by jayunit100** [comment]
we have  service loadbalancing tests in the e2es triggered above: 

```
- Kubernetes e2e suite: [sig-windows] Services should be able to create a functioning NodePort service for Windows
- Kubernetes e2e suite: [sig-network] Services should be able to change the type from ExternalName to ClusterIP [Conformance]
- Services should serve multiport endpoints from pods
```

**Comment by jayunit100** [comment]
since the tests are passing, i think we should just merge and watch it, iterate if necessar

/lgtm
/approve

**Comment by anfernee** [comment]
> imo we could merge this but , quick question... this is tricky to reason about....
> 
> whats the urgency on merging this patch ? we could...
> 
> 1. merge it now, watch it closely, and iterate where needed OR, we could ...
> 2. do a test coverage based approach
> 
> * along with this patch,  add a `sig-windows` tagged test that ensures that `externalTrafficPolicy=local` when implemented, always sends out correct health checks?  i think it would be doable ?
> * otherwise, if you file an issue in https://github.com/K8sbykeshed/k8s-service-lb-validator/, our new service lb tests we can try to create this test for you .
> 
> side note - I THINK we SKIP the ESIPP tests for IPVS... so it seems like the overall impl externalTrafficPolicy might not be uniform in kube proxy
> 
> anyways lmk what you think is best Yongkun Anfernee Gui ! if antonio and rob are on board, then merge+iterate seems like an ok idea to me.... i dont fully understand the kube proxy health check parts so ill have to learn more about it today as i read this.

since default behavior is unchanged, this should be safe to merge. In terms of whether this fix pass-through load balancer, we need to turn on those 2 options and find a pass-through lb (Google's GCLB) to check. Since this is very provider specific. Where do you recommend the test should be living in?

**Comment by marosset** [comment]
/assign @thockin
For final review (or re-assignment)

**Comment by anfernee** [comment]
Thanks everyone for all the help for this PR. Finally almost there. Special thanks for the original author of this PR @jeremyje

**Comment by anfernee** [comment]
When can we merge this PR?

**Comment by jayunit100** [comment]
Agree it's fine -> Let's merge and keep an eye out

**Comment by jayunit100** [comment]
Double checking that you got the earlier note: we should make sure our externalTrafficPolicy.... tests are covering stuff like this somehow, eventually, on the sig-win side

**Comment by anfernee** [comment]
> Double checking that you got the earlier note: we should make sure our externalTrafficPolicy.... tests are covering stuff like this somehow, eventually, on the sig-win side

I am not exactly familiar with the test setup but I'll check it out. If the test runs on GCE, then we can enable the option there. Existing e2e service test should be able to catch it.

**Comment by anfernee** [comment]
Fixed `pkg/generated/openapi/zz_generated.openapi.go`. Pushed again.

**Comment by anfernee** [comment]
This PR requires LGTM again @jayunit100 @jsturtevant

**Pull request merged**
